[Clang] Implement CWG2369 "Ordering between constraints and substitution"

zyn0217 opened the pull request:
This patch partially implements CWG 2369 for non-lambda-constrained functions.

Lambdas are left intact at this point because we need extra work to correctly instantiate captures before the function instantiation.

As a premise of CWG2369, this patch also implements CWG2770 to ensure the function parameters are instantiated on demand.

Closes https://github.com/llvm/llvm-project/issues/54440

zyn0217 committed:
CWG 2369

zyn0217 committed:
Rectify diagnostics

zyn0217 committed:
[Clang][NFCI] Slightly refactor getTemplateInstantiationArgs()

zyn0217 committed:
Instantiate function parameters as needed

zyn0217 committed:
Merge branch 'main' into cwg-2369-2

zyn0217 committed:
Resolve a merge conflict

zyn0217 committed:
Remove debugging logs

zyn0217 committed:
Revert unrelated changes

zyn0217 committed:
Fix tests

zyn0217 committed:
Fix libcxx tests

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1122,6 +1122,154 @@ bool Sema::CheckInstantiatedFunctionTemplateConstraints(
+    if (auto *PVD = dyn_cast_if_present<ParmVarDecl>(D);
+        PVD && PVD->getDeclContext() == PrimaryTemplatedFunction &&
+        !InstantiatedDecls.contains(PVD)) {
+      instantiateParameterToScope(PVD, *SemaRef.CurrentInstantiationScope);
```
We probably need to handle errors here.
But I think we should instantiate _ALL_ the parameters, not just the one referenced in the constraint

zyn0217 replied:
That would cause an implementation divergence with GCC: https://gcc.godbolt.org/z/7MobE9b61
If we instantiate all parameters, we wouldn't accept such a case that GCC currently accepts.

The standard seems *vague* ([It](https://cplusplus.github.io/CWG/issues/2369.html) just said `function type` shouldn't be replaced before the constraint evaluation) about this, but I presume the intent was to instantiate parameters on demand.

cor3ntin replied:
I did some research, this is https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#2770

cor3ntin replied:
See also https://github.com/cplusplus/CWG/issues/364 
https://cplusplus.github.io/CWG/issues/2769.html

https://gcc.godbolt.org/z/WrfG5Gxaj

I think instead of doing the substitution there, we want to do it where we check atomic constraints

cor3ntin replied:
@zygoloid

zyn0217 replied:
> I think instead of doing the substitution there, we want to do it where we check atomic constraints

Err.. I just tried and I realized there will be much more complexity then:
IIUC we need to move the substitution to `TemplateInstantiator` which just delegates `TryExpandParameterPacks` to `Sema.CheckParameterPacksForExpansion`, so for a pack expansion, e.g.

```cpp
void f18(auto... x)
  requires(sizeof...(x) == 2);

void foo() {
  f18('c');
}
```

We would end up calling `CurrentInstantiationScope->findInstantiationOf` inside `CheckParameterPacksForExpansion` when trying to transform the `sizeof...` expression. Unfortunately we don't have a chance to instantiate `x` between the transform of the expression starts and the `findInstantiation` call, therefore we would crash.

One alternative is to duplicate much of the `CheckParameterPacksForExpansion` to `TryExpandParameterPacks`, and we instantiate the Decls there. But I feel that would clutter up everything and we probably couldn't accept the maintenance cost.

zyn0217 replied:
Hmmm, never mind. Probably things are not so dreadful than I thought.

zyn0217 committed:
Instantiate the decls as we substituting constraints

mordante committed:
[libc++][chono] Use hidden friends for leap_second comparison.

The function

    template<class Duration>
      requires three_way_comparable_with<sys_seconds, sys_time<Duration>>
      constexpr auto operator<=>(const leap_second& x, const sys_time<Duration>& y) noexcept;

Has a recursive constrained. This caused an infinite loop in GCC and is
now hit by https://github.com/llvm/llvm-project/pull/102857.

A fix would be to make this function a hidden friend, this solution is
propsed in LWG4139.

For consistency all comparisons are made hidden friends.
Since the issue causes compilation failures no additional test are
needed.

zyn0217 committed:
clang-format

zyn0217 commented:
@mordante 
I saw some tests are "unexpectedly passed" on ARM platforms: https://buildkite.com/llvm-project/libcxx-ci/builds/37165#0191685f-c770-41d1-a8a4-8819da1b1802
Is that something undesired / why did we mark them XFAIL previously?

zyn0217 committed:
Merge branch 'main' into cwg-2369

zyn0217 committed:
Merge branch 'main' into cwg-2369

zyn0217 committed:
Quote the standard wordings, update the DR status page

zyn0217 committed:
Revert unrelated changes

zyn0217 committed:
oops, I forgot it

Endilll commented on clang/test/CXX/drs/cwg27xx.cpp:
```diff
@@ -101,6 +101,26 @@ static_assert(!__is_layout_compatible(StructWithAnonUnion, StructWithAnonUnion3)
+};
+
+template<typename T>
+int f(T t, typename B<T>::type u) requires (sizeof t == 1);
```
Using a `sizeof expression` form (without parentheses) seems rather peculiar to me. Is there a reason for this?

zyn0217 replied:
No, @cor3ntin provides me with it in the above comment, and I just copy-pasted it.

Sirraide commented:
(Unassigning myself because I’m not familiar enough w/ the details of substitution to review this in a meaningful way).

That said, I think this is still missing a release note isn’t it? :)

zyn0217 commented:
> (Unassigning myself because I’m not familiar enough w/ the details of substitution to review this in a meaningful way).

No problem :)

> That said, I think this is still missing a release note isn’t it? :)

Yeah, I realized it the moment after I pushed the last commit, and I decided to leave that in the next push so I don't make commits piecemeal.

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -13051,11 +13051,14 @@ class Sema final : public SemaBase {
   /// \param DC In the event we don't HAVE a declaration yet, we instead provide
-  ///  the decl context where it will be created.  In this case, the `Innermost`
-  ///  should likely be provided.  If ND is non-null, this is ignored.
+  ///  the decl context where it will be created.  In this case, the \p
+  ///  Innermost should likely be provided.  If \p ND is non-null and \p
```
likely?

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -13051,11 +13051,14 @@ class Sema final : public SemaBase {
   ///
   /// \param Innermost if non-NULL, specifies a template argument list for the
-  /// template declaration passed as ND.
+  /// template declaration passed as \p ND. The next declaration context would
+  /// be switched to \p DC if present; otherwise, it would be the semantic
```
I am not sure what "Next Declaration Context" refers too here

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1144,6 +1164,46 @@ bool Sema::CheckInstantiatedFunctionTemplateConstraints(
                                      PointOfInstantiation, Satisfaction);
 }
 
+bool Sema::CheckFunctionConstraintsWithoutInstantiation(
```
This is very similar to `CheckInstantiatedFunctionTemplateConstraints` can you either reuse that or share some code?

zyn0217 replied:
I moved it into CheckFunctionTemplateConstraints() - calls to e.g. SetupConstraintCheckingTemplateArgumentsAndScope() inside couldn't be reused very simply so I added another internal function to handle MLTALs.

cor3ntin commented on clang/lib/Sema/SemaTemplate.cpp:
```diff
@@ -5585,11 +5585,12 @@ bool Sema::CheckTemplateArgumentList(
-        /*RelativeToPrimary=*/true,
-        /*Pattern=*/nullptr,
-        /*ForConceptInstantiation=*/true);
+    MultiLevelTemplateArgumentList MLTAL =
+        getTemplateInstantiationArgs(Template, Template->getDeclContext(),
+                                     /*Final=*/false, CanonicalConverted,
+                                     /*RelativeToPrimary=*/true,
+                                     /*Pattern=*/nullptr,
+                                     /*ForConceptInstantiation=*/true);
```
Can you explain why we do not use `NewContext` anymore?

zyn0217 replied:
The `NewContext` argument didn't actually make any difference to `getTemplateInstantiationArgs()` previously, as it would always move to `Template->getDeclContext()` for the next traversal if the Innermost parameter (which is `CanonicalConverted`) was present. So I changed it to `Template->getDeclContext()` to clarify the behavior.

However, I can revert the changes on `getTemplateInstantiationArgs()` after @sdkrystian's refactoring patch, so this is left only for the time being.

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -3876,6 +3864,41 @@ TemplateDeductionResult Sema::FinishTemplateArgumentDeduction(
     Owner = FD->getLexicalDeclContext();
   }
+
+  // C++20 [temp.deduct.general]p5: [DR2369]
```
```suggestion
  // C++20 [temp.deduct.general]p5 (CWG2369):
```

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -3876,6 +3864,41 @@ TemplateDeductionResult Sema::FinishTemplateArgumentDeduction(
+      !PartialOverloading ||
+      CanonicalBuilder.size() ==
+          FunctionTemplate->getTemplateParameters()->size();
+  // FIXME: We haven't implemented DR2369 for lambdas yet, because we need
```
```suggestion
  // FIXME: We haven't implemented CWG2369 for lambdas yet, because we need
```

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -3876,6 +3864,41 @@ TemplateDeductionResult Sema::FinishTemplateArgumentDeduction(
+      return TemplateDeductionResult::ConstraintsNotSatisfied;
+    }
+  }
+  // C++ [temp.deduct.call]p10: [DR1391]
```
```suggestion
  // C++ [temp.deduct.call]p10 (CWG1391):
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1397,12 +1415,25 @@ namespace {
+              PVD &&
+              PVD->getDeclContext() ==
+                  FunctionDCForParameterDeclInstantiation &&
+              !ParameterInstantiationScope->findInstantiationUnsafe(PVD)) {
+            // Fall through to the default lookup even if we have failed to
+            // instantiate anything. We're likely to crash thereafter.
+            instantiateParameterToScope(PVD, *ParameterInstantiationScope);
```
Crashing doesn't see appropriate. Can that ever fail? Should we add an assert?

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1841,9 +1872,71 @@ Decl *TemplateInstantiator::TransformDecl(SourceLocation Loc, Decl *D) {
   return SemaRef.FindInstantiatedDecl(Loc, cast<NamedDecl>(D), TemplateArgs);
 }
 
+bool TemplateInstantiator::instantiateParameterToScope(
```
Can we reuse the logic of `SetupConstraintScope` / `addInstantiatedParametersToScope` ? It seems very similar

cor3ntin reviewed (commented):
Thanks!
I did a first pass on the review, I have a few questions

mizvekov commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1144,6 +1164,46 @@ bool Sema::CheckInstantiatedFunctionTemplateConstraints(
+
+  // Enter the scope of this instantiation. We don't use
+  // PushDeclContext because we don't have a scope.
+  LocalInstantiationScope Scope(*this);
```
I am not sure how `PushDeclContext` is related to the instantiation scope.

`PushDeclContext` would have changed the CurContext. If you need that, you can use `ContextRAII`

zyn0217 replied:
Removed. This comment was copied elsewhere, and I think the very original meaning was that "we don't have a parser Scope so we have to use something else."

mizvekov commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1144,6 +1164,46 @@ bool Sema::CheckInstantiatedFunctionTemplateConstraints(
+                                   /*Pattern=*/nullptr,
+                                   /*ForConstraintInstantiation=*/true);
+
+  Qualifiers ThisQuals;
+  CXXRecordDecl *Record = nullptr;
+  if (auto *Method = dyn_cast<CXXMethodDecl>(FD)) {
+    ThisQuals = Method->getMethodQualifiers();
+    Record = Method->getParent();
+  }
+  CXXThisScopeRAII ThisScope(*this, Record, ThisQuals, Record != nullptr);
```
```suggestion
  std::optional<CXXThisScopeRAII> ThisScope;
  if (auto *Method = dyn_cast<CXXMethodDecl>(FD))
    ThisSope.emplace(*this,  Method->getParent(), Method->getMethodQualifiers());
```

zyn0217 committed:
Merge branch 'main' into cwg-2369

zyn0217 committed:
Simplify the implementation

zyn0217 committed:
Add parentheses to sizeof expression

zyn0217 committed:
Merge branch 'main' into cwg-2369

zyn0217 committed:
Revert comment changes

zyn0217 committed:
Rebase on top of the new getTemplateInstantiationArgs()

zyn0217 committed:
Merge branch 'main' into cwg-2369

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -1916,9 +1935,33 @@ Decl *TemplateInstantiator::TransformDecl(SourceLocation Loc, Decl *D) {
+  if (SemaRef.CurrentInstantiationScope->findInstantiationUnsafe(OldParm))
+    return false;
+  // The current context might have been changed in the process of transforming
+  // lambda expression. So resume it before we substitute into the parameter.
```
```suggestion
  // lambda expression. So restore it before we substitute into the parameter.
```

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -4042,8 +4062,8 @@ TemplateDeductionResult Sema::FinishTemplateArgumentDeduction(
   //   deduction fails.
-  if (!IsIncomplete) {
-    if (CheckInstantiatedFunctionTemplateConstraints(
+  if (IsLambda && !IsIncomplete) {
```
can you add a fixme comment here?

zyn0217 replied:
There's one on line 4003, maybe that just suffices?

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -4008,6 +3996,38 @@ TemplateDeductionResult Sema::FinishTemplateArgumentDeduction(
+      return TemplateDeductionResult::ConstraintsNotSatisfied;
+    }
+  }
+  // C++ [temp.deduct.call]p10: [DR1391]
```
```suggestion
  // CWG1391
```

zyn0217 replied:
Oops, I thought I have fixed it everywhere...

cor3ntin commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -4008,6 +3996,38 @@ TemplateDeductionResult Sema::FinishTemplateArgumentDeduction(
+            CanonicalBuilder, Info.AssociatedConstraintsSatisfaction))
+      return TemplateDeductionResult::MiscellaneousDeductionFailure;
+    if (!Info.AssociatedConstraintsSatisfaction.IsSatisfied) {
+      Info.reset(Info.takeSugared(),
+                 TemplateArgumentList::CreateCopy(Context, CanonicalBuilder));
```
Why is that necessary?

zyn0217 replied:
(Frankly these lines are also copied from the previous failure handling)

I *guess* the instantiated concept-related nodes that were created during the checking would use the TemplateArgumentLists created on the ASTContext, or rather they would take ownership of the TemplateArgumentList. So, in order for the pilfered argument lists to work for other clients of TemplateDeductionInfo, we have to make a copy of the list.

(I saw the patch that tried to preserve sugars in concepts turn this line into "copy sugared arguments, but do nothing for canonical arguments")

But I'm not entirely sure of such deduction things, so probably @mizvekov would correct my understanding in some way.

cor3ntin replied:
ping @mizvekov

mizvekov replied:
Yeah, what @zyn0217 said is correct, the canonical argument list will have references persist in AST nodes, so it needs to be copied here.

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1517,7 +1555,7 @@ substituteParameterMappings(Sema &S, NormalizedConstraint &N,
                                         const ConceptSpecializationExpr *CSE) {
   MultiLevelTemplateArgumentList MLTAL = S.getTemplateInstantiationArgs(
-      CSE->getNamedConcept(), CSE->getNamedConcept()->getLexicalDeclContext(),
+      CSE->getNamedConcept(), CSE->getNamedConcept()->getDeclContext(),
```
Do you have tests with friend functions?

zyn0217 replied:
This is another leftover of my previous changes on `getTemplateInstantiationArgs()`... Will revert it

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1109,12 +1109,50 @@ bool Sema::EnsureTemplateArgumentListConstraints(
+                                           /*ForConstraintInstantiation=*/true);
+
+  std::optional<Sema::CXXThisScopeRAII> ThisScope;
+  if (auto *Method = dyn_cast<CXXMethodDecl>(FD))
+    ThisScope.emplace(SemaRef, /*Record=*/Method->getParent(),
+                      /*ThisQuals=*/Method->getMethodQualifiers());
```
When is that useful?

zyn0217 replied:
Honestly, I don't know. :(

I copied that from the previous implementation, and I was expecting it to make a difference in situations where member function calls are involved in a constraint evaluation. But I ran the regression tests with that removed and everything seemed fine.

I will have another try on a build without this patch. It could also be the case we're missing some tests, but I don't know...

zyn0217 replied:
Huh, I removed these `CXXThisScopeRAII` lines in `CheckInstantiatedFunctionTemplateConstraints()` and all libcxx and clang regressions still passed. I guess we probably could remove these setups there

cor3ntin replied:
maybe try to play with things like `requires(__is_same_as(s, decltype(*this)));`

zyn0217 replied:
I tried to contrive a case like

```cpp
template <class T>
struct S {
  void foo() requires(__is_same_as(S&, decltype(*this))) {}
};

void g() {
  S<int>().foo();
}
```

But there would still be no difference if we did the `CXXThisScopeRAII` things away. (It always compiles.)

Maybe I'm running out of ideas right now, so let's just preserve these "strange but plausible" things?

mizvekov commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1109,12 +1109,50 @@ bool Sema::EnsureTemplateArgumentListConstraints(
+                                           /*RelativeToPrimary=*/true,
+                                           /*ForConstraintInstantiation=*/true);
+
+  std::optional<Sema::CXXThisScopeRAII> ThisScope;
```
Do we need to point CurContext to the function here as well?

zyn0217 replied:
Yes, I think so - but I couldn't find a case where that affects anything.

cor3ntin replied:
Better safe than sorry!

zyn0217 replied:
Added a ContextRAII anyway

zyn0217 committed:
Merge branch 'main' into cwg-2369

zyn0217 committed:
Address a few comments

cor3ntin commented on clang/include/clang/Sema/Template.h:
```diff
@@ -522,6 +522,9 @@ enum class TemplateSubstitutionKind : char {
     llvm::PointerUnion<Decl *, DeclArgumentPack *> *
     findInstantiationOf(const Decl *D);
 
+    llvm::PointerUnion<Decl *, DeclArgumentPack *> *
+    findInstantiationUnsafe(const Decl *D);
```
Can you add a comment explaining what this does?

zyn0217 committed:
Merge branch 'main' into cwg-2369

zyn0217 committed:
Address more comments

zyn0217 committed:
Align the comment formats in FinishTemplateArgumentDeduction()

zyn0217 commented:
Sadly I think I should figure out a way to remove reliance on the refactoring work of `getTemplateInstantiationArgs()`, as it is still unstable at this point.

cor3ntin commented:
@zyn0217 should we wait then?

zyn0217 committed:
Merge branch 'main' into cwg-2369

zyn0217 committed:
Decouple from the new version of getTemplateInstantiationArgs()

zyn0217 commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -13044,6 +13045,17 @@ class Sema final : public SemaBase {
       bool SkipForSpecialization = false,
       bool ForDefaultArgumentSubstitution = false);
 
+  /// Apart from storing the result to \p Result, this behaves the same as
+  /// another overload.
+  void getTemplateInstantiationArgs(
+      MultiLevelTemplateArgumentList &Result, const NamedDecl *D,
+      const DeclContext *DC = nullptr, bool Final = false,
+      std::optional<ArrayRef<TemplateArgument>> Innermost = std::nullopt,
+      bool RelativeToPrimary = false, const FunctionDecl *Pattern = nullptr,
+      bool ForConstraintInstantiation = false,
+      bool SkipForSpecialization = false,
+      bool ForDefaultArgumentSubstitution = false);
+
```
This was taken from #110387, which has also been reverted due to its dependency on the refactoring work. 

The extra parameters can certainly be removed alongside the relanding of that huge patch.

cor3ntin replied:
We can remove that now

zyn0217 replied:
(For clarity, this function doesn't insert an empty list for FunctionDecl that is not a specialization. So we need to do that right before the call)

zyn0217 commented:
@cor3ntin Not much, we just need to move the handling of the friend declaration outside of `getTemplateInstantiationArgs()` :)

Do you see any other outstanding issues? If not, can we go ahead with the status quo so we have sufficient timeframe in clang 20 cycle in case something else goes wrong?

cor3ntin commented:
I think I'd be happy with that.
@mizvekov You have an opinion?

erichkeane commented:
> @cor3ntin Not much, we just need to move the handling of the friend declaration outside of `getTemplateInstantiationArgs()` :)
> 
> Do you see any other outstanding issues? If not, can we go ahead with the status quo so we have sufficient timeframe in clang 20 cycle in case something else goes wrong?

I don't think I'd be a fan of removing the friend declaration outside of `getTemplateInstantiationArgs`.  That is supposed to be the 'single source' of truth for that, and should work on just about any declaration that could potentially be templated.  So I think we need to do the work to improve it isntead.

sdkrystian commented:
FWIW, I'm going to try reland the `getTemplateInstantiationArgs` refactoring again this week.

zyn0217 commented:
> > @cor3ntin Not much, we just need to move the handling of the friend declaration outside of `getTemplateInstantiationArgs()` :)
> 
> > 
> 
> > Do you see any other outstanding issues? If not, can we go ahead with the status quo so we have sufficient timeframe in clang 20 cycle in case something else goes wrong?
> 
> 
> 
> I don't think I'd be a fan of removing the friend declaration outside of `getTemplateInstantiationArgs`.  That is supposed to be the 'single source' of truth for that, and should work on just about any declaration that could potentially be templated.  So I think we need to do the work to improve it isntead.


Yeah, I admit that the current approach is suboptimal, as earlier iterations demonstrated that the patch could be cleanly built on top of the getTemplateInstantiationArgs() refactoring, without requiring any special-casing outside of that function.

So I doubt if there’s much value in improving the old version of the function, especially since the refactoring work can be reapplied at any time.

cor3ntin commented:
Lets wait for @sdkrystian's pr

zyn0217 commented:
@cor3ntin Do we still want to block on the getTemplateInstantiationArgs() stuff or move forward anyway so we can (hopefully) have this in the upcoming clang 20? FWIW, this patch currently doesn't rely on the refactoring, using only a ternary operand as a workaround for which it should have been handled by getTemplateInstantiationArgs().

cor3ntin commented:
Yeah, I was hoping that PR would materialize sooner. I agree we should land this soon so let's not wait any longer.

zyn0217 committed:
Merge branch 'main' into cwg-2369

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -6089,16 +6090,16 @@ bool TreeTransform<Derived>::TransformFunctionTypeParams(
-    Sema::ExtParameterInfoBuilder &PInfos,
-    unsigned *LastParamTransformed) {
+    Sema::ExtParameterInfoBuilder &PInfos, unsigned *LastParamTransformed,
+    bool IgnoreParameterIndex) {
```
This is just used in `assert` ?  should we mark it maybe_unused?

zyn0217 replied:
Yes, and the assertion feels awkward in this case because we only instantiate what we use, which is something we’ve never done with function parameters.

I don’t recall seeing any warnings, even for my local release build, but my memory on that is a bit unclear. I’ll look into it soon.

zyn0217 replied:
... Or let me see if I can refactor some common logics from TransformFunctionTypeParams

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1111,12 +1111,57 @@ bool Sema::EnsureTemplateArgumentListConstraints(
+
+  // FIXME: Add TemplateArgs through the 'Innermost' parameter once
+  // the refactoring of getTemplateInstantiationArgs() relands.
+  MultiLevelTemplateArgumentList MLTAL;
+  MLTAL.addOuterTemplateArguments(Template, /*Args=*/TemplateArgs,
```
I am a bit confused how this work. If there are multiple levels of template, this will end up being the wrong depth, would it not?

zyn0217 replied:
We would collect outer template arguments in the followed getTemplateInstantiationArgs() call, which is adapted to store them in MLTAL. So the depths should remain relative to the primary, which is what we expected?

cor3ntin replied:
Hum, nevermind, I forgot getTemplateInstantiationArgs walks upwards. 
Remind me what stops us to use `Innermost` here?

cor3ntin replied:
Ok, so we need to set `Template` as the innermost associated decl - and use FD's decl context elsewhere
Does it work if we pass an empty array to getTemplateInstantiationArgs()
and subsequently call replaceInnermostTemplateArguments with the correct arguments/decl?

zyn0217 replied:
This sounds hacky but I like it! I'll run the test locally and see if it works

zyn0217 replied:
Yes, at least check-clang with assertion passed! Thank you @cor3ntin :)

We could really land this much earlier than now if I had come up with this idea sooner 😭

zyn0217 committed:
Don't call TransformFunctionTypeParams; expand the parameters ourselves

zyn0217 committed:
Remove unnecessary code

zyn0217 committed:
Use MLTAL.replaceInnermostTemplateArguments to avoid the special-casing

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1111,12 +1111,55 @@ bool Sema::EnsureTemplateArgumentListConstraints(
+
+  // FIXME: Add TemplateArgs through the 'Innermost' parameter once
+  // the refactoring of getTemplateInstantiationArgs() relands.
+  MultiLevelTemplateArgumentList MLTAL;
+  MLTAL.addOuterTemplateArguments(Template, std::nullopt, /*Final=*/false);
+  SemaRef.getTemplateInstantiationArgs(
+      MLTAL, /*D=*/FD, FD,
+      /*Final=*/false, /*Innermost=*/std::nullopt, /*RelativeToPrimary=*/true,
+      /*Pattern=*/nullptr, /*ForConstraintInstantiation=*/true);
+  MLTAL.replaceInnermostTemplateArguments(Template, TemplateArgs);
```
```suggestion
  MultiLevelTemplateArgumentList MLTAL = SemaRef.getTemplateInstantiationArgs(
      /*D=*/FD, FD,
      /*Final=*/false, /*Innermost=/*{}, /*RelativeToPrimary=*/true,
      /*Pattern=*/nullptr, /*ForConstraintInstantiation=*/true);
  MLTAL.replaceInnermostTemplateArguments(Template, TemplateArgs);
```

zyn0217 replied:
* See my comment below - we can't use that overload :)

cor3ntin replied:
I'm very confused. 
I am saying that we can pass an empty array to innermost, which creates a dummy inner most list with an incorrect declaration - but then call replaceInnermostTemplateArguments which then set the right list/declaration.

Why doesn't that work?

zyn0217 replied:
`replaceInnermostTemplateArguments()` has an assertion that the updated AssociatedDecl must match the previously added one. So I'm afraid an 'incorrect' declaration as a placeholder wouldn't work here

Even if we could bypass that restriction, we still wouldn't be able to navigate to the desired lexical declaration context after the inner argument list.

https://github.com/llvm/llvm-project/blob/ee9be864bcc5e3cc89f5f23485db2285ad7119f7/clang/lib/Sema/SemaTemplateInstantiate.cpp#L485-L498

(Where Response::UseNextDecl() is always the semantic DC regardless of the friendness.)

https://github.com/llvm/llvm-project/blob/ee9be864bcc5e3cc89f5f23485db2285ad7119f7/clang/lib/Sema/SemaTemplateInstantiate.cpp#L74-L76)

cor3ntin replied:
The assert is what I was missing... it's a very weird assert

```cpp
assert((TemplateArgumentLists[0].AssociatedDeclAndFinal.getPointer() ||
                TemplateArgumentLists[0].AssociatedDeclAndFinal.getPointer() ==
                    AssociatedDecl) &&
               "Trying to change incorrect declaration?");```
```

Does it actually fire?
I'd be fine with temporarily removing it

zyn0217 replied:
Then we still face the question of which declaration to visit after Innermost. Without Krystian's patch, I'm afraid we would also need to modify the function.

The reason I added another overload is that it's actually useful for @mizvekov to reapply #110387, which also depends on that work.

Additionally, I don’t think it would be a significant challenge to reland the refactoring work - just aligning the parameters should be fine. I’d definitely be happy to revisit/fix the workaround here thereafter.

cor3ntin reviewed (approved):
So, I'll will approve that we need to land the issue resolution in Clang 20, but we need to make sure we cleanup `getTemplateInstantiationArgs` soon :)

Thanks for working through this non-trivial PR!

zyn0217 commented:
Thanks so much for the review! I'll keep an eye on that PR, and there’s some post-work to do: 1. Cleanup 2. Adaptation for lambdas

zyn0217 commented on clang/test/CXX/drs/cwg27xx.cpp:
```diff
@@ -174,6 +174,26 @@ static_assert(!__is_layout_compatible(StructWithAnonUnion, StructWithAnonUnion3)
 } // namespace cwg2759
 
+#if __cplusplus >= 202002L
+namespace cwg2770 { // cwg2770: 20
```
Yes, I implemented what that issue proposed.

I saw the script removing the entire cwg2770 line after I added this test. Could you instruct me on what to do with it? Should we leave out the test until the issue gets resolved?

Endilll replied:
> Yes, I implemented what that issue proposed.

Good to know! I fixed it with NFC commits myself, no worries.

> I saw the script removing the entire cwg2770 line after I added this test.

What happened is that script has printed `Issue 2770 is marked '', which differs from CWG index status 'open'`, and stopped right there, before emitting HTML for CWG2770. I guess in order to clear the confusion, I should update the script to not write anything to file until it's sure everything is good.

> Could you instruct me on what to do with it? Should we leave out the test until the issue gets resolved?

No, the script is able to handle tests for not yet resolved issues. That's how you do it:
https://github.com/llvm/llvm-project/blob/1229b78f1676d56d0d386a5c4cde8f98a78d0989/clang/test/CXX/drs/cwg27xx.cpp#L177

I also moved CWG2369 test from the end of `cwg23xx.cpp` right before CWG2370 test, because that's how DR tests are ordered.

zyn0217 replied:
Thanks for the post cleanup! Just learned how to deal with these unresolved issues in the future :)

mysterymath commented:
We've started seeing libc++ test failures in the Fuchsia toolchain buildbots for Windows: https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-windows-x64/b8726615170786362641/overview

This may be related to the discussion in #121881, but the error looks slightly different.

This is definitely the only plausible PR in the blamelist: https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-windows-x64/b8726615170786362641/blamelist
```
llvm-libc++-static-clangcl.cfg.in :: std/utilities/optional/optional.relops/less_than.pass.cpp
==============
# | In file included from C:\b\s\w\ir\x\w\llvm-llvm-project\libcxx\test\std\utilities\optional\optional.relops\less_than.pass.cpp:14:
# | C:/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-pc-windows-msvc-bins/libcxx/test-suite-install/include/c++/v1\optional:1247:1: error: cannot mangle this dependent name type yet
# |  1247 | _LIBCPP_HIDE_FROM_ABI constexpr compare_three_way_result_t<_Tp, _Up>
# |       | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |  1248 | operator<=>(const optional<_Tp>& __x, const _Up& __v) {
# |       | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |  1249 |   return __x.has_value() ? *__x <=> __v : strong_ordering::less;
# |       |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# |  1250 | }
# |       | ~
```

zmodem commented:
> We've started seeing libc++ test failures in the Fuchsia toolchain buildbots for Windows
 
We're hitting the same error in Chromium builds. Here is a stand-alone reproducer: https://crbug.com/388428503#comment4

cor3ntin commented:
> > We've started seeing libc++ test failures in the Fuchsia toolchain buildbots for Windows
> 
> We're hitting the same error in Chromium builds. Here is a stand-alone reproducer: https://crbug.com/388428503#comment4

Thanks. @Endilll is working on a reduction, so that we get a better sense of what the issue is

zyn0217 commented:
@cor3ntin Do we need to revert it for now?

cor3ntin commented:
> @cor3ntin Do we need to revert it for now?

I'd like us to figure out if it's a conformance issue first - but bearing resolution we should revert before end of week imo

zmodem commented:
Here's what I got from creduce:

```
$ cat /tmp/a.ii
template <class, class _Up>
using compare_three_way_result_t = _Up ::type;

struct __sfinae_assign_base {};

template <class _Tp>
concept __is_derived_from_optional =
    requires(_Tp __t) { []<class _Up>(_Up) {}(__t); };

template <class _Tp, class _Up>
  requires(__is_derived_from_optional<_Up>)
compare_three_way_result_t<_Tp, _Up> operator<=>(_Tp, _Up);

struct RuntimeModeArgs {
  auto operator<=>(const RuntimeModeArgs &) const = default;
  __sfinae_assign_base needs_admin;
};

$ build/bin/clang.bad -cc1 -triple x86_64-pc-windows-msvc19.34.0 -emit-obj -fms-extensions -fms-compatibility -std=c++20 -w /tmp/a.ii
/tmp/a.ii:12:1: error: cannot mangle this dependent name type yet
   12 | compare_three_way_result_t<_Tp, _Up> operator<=>(_Tp, _Up);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

zyn0217 commented:
I see the problem here:

When substituting into the constraint __is_derived_from_optional with [_Tp = __sfinae_assign_base] prior to substituting the three-way operator, we must also substitute into the lambda call expression. This requires instantiating the lambda body to deduce its return type, and instantiating the lambda body necessitates mangling its parent for Microsoft mangling. In this case, the lambda resides within the function declaration `compare_three_way_result_t<_Tp, _Up> operator<=>(_Tp, _Up)`, which remains uninstantiated at this point!

@cor3ntin

cor3ntin commented:
@zyn0217 Ouch. Lets revert and try to figure that out...  @erichkeane

erichkeane commented:
Oof, hrmph.  It seems to me that we just have to be slightly more aggressive at making sure we instantiate stuff I think.

This sentence:
`In this case, the lambda resides within the function declaration compare_three_way_result_t<_Tp, _Up> operator<=>(_Tp, _Up), which remains uninstantiated at this point!`

Is kinda wrong, it should be in the instantiation of that function...

zyn0217 commented:
> Is kinda wrong, it should be in the instantiation of that function...

@erichkeane I don’t quite understand: the CWG issue itself requires us to check the constraint before substituting into the function during template argument deduction. So it shouldn’t live in a specialization because we don’t really have one. This situation feels more like a chicken-and-egg problem.

GCC doesn’t encounter this issue because it doesn’t need to support the MSVC ABI. The problem only arises with the MSVC ABI because it requires parent mangling when instantiating the lambda body.

frederick-vs-ja commented:
Do we want to add some workaround to libc++'s `<optional>`?

cor3ntin commented:
@mizvekov can you investigate whether #107942 helps with this? Thanks

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -1111,12 +1111,55 @@ bool Sema::EnsureTemplateArgumentListConstraints(
+      /*Final=*/false, /*Innermost=*/std::nullopt, /*RelativeToPrimary=*/true,
+      /*Pattern=*/nullptr, /*ForConstraintInstantiation=*/true);
+  MLTAL.replaceInnermostTemplateArguments(Template, TemplateArgs);
+
```
I think we want to enter an Unevaluated `ExpressionEvaluationContext`
Given the windows bug, I think the `SavedContext` might not be correct. Can you play with that?

erichkeane commented:
> > Is kinda wrong, it should be in the instantiation of that function...
> 
> @erichkeane I don’t quite understand: the CWG issue itself requires us to check the constraint before substituting into the function during template argument deduction. So it shouldn’t live in a specialization because we don’t really have one. This situation feels more like a chicken-and-egg problem.
> 
> GCC doesn’t encounter this issue because it doesn’t need to support the MSVC ABI. The problem only arises with the MSVC ABI because it requires parent mangling when instantiating the lambda body.

Actually, I perhaps misread the example (I think I inserted extra curleys there).  It isn't clear to me why we are mangling the name of the lambda.  It isn't in a context where it should be emitted?  Corentin might be right about the starting an unevaluated context.

It also isn't clear to me why compare_three_way_result_t is being instantiated with a lambda in it, but the example is one that perhaps needs more contemplating.

cor3ntin commented:
I spent sometimes on this with @zyn0217 

```cpp

template <class, class _Up>
using compare_three_way_result_t = _Up ::type;

struct __sfinae_assign_base {};

template <class _Tp, class _Up>
  requires ([]<class W>(W) { return true; }(_Up())) // #1
compare_three_way_result_t<_Tp, _Up> operator<=>(_Tp, _Up); //#2

struct RuntimeModeArgs {
  auto operator<=>(const RuntimeModeArgs &) const = default;
  __sfinae_assign_base needs_admin;
};
```

In `#1`, the lambda has an `auto` return type. During constraint satisfaction checking (*before* `#2` gets specialized), we need to call `DeduceReturnType` which calls `InstantiateFunctionDefinition`. 

In turn `InstantiateFunctionDefinition` ends up calling `EmitGlobalDecl` through:
https://github.com/llvm/llvm-project/blob/81ae6686ab026c820f2f23d3c18cdd65ecb2ae7c/clang/lib/Sema/SemaTemplateInstantiateDecl.cpp#L5288

This in turns requires the lambda to have a mangled name, causing its enclosing, dependent function template type to get mangled on the windows abi, which obviously isn't going to work. 

Fortunately, we do not need to mangle constraints, so we just need to find a reasonable way to skip the codegen.
It's weird that we try to emit anything from an unevaluated context to begin with.


This patch does the trick by skipping `EmitGlobalDecl`  from unevaluated and immediate contexts

```diff
diff --git a/clang/include/clang/Sema/Sema.h b/clang/include/clang/Sema/Sema.h
index 18fd95f77ec2..0e30f21ab68b 100644
--- a/clang/include/clang/Sema/Sema.h
+++ b/clang/include/clang/Sema/Sema.h
@@ -13721,7 +13721,8 @@ public:
                                      FunctionDecl *Function,
                                      bool Recursive = false,
                                      bool DefinitionRequired = false,
-                                     bool AtEndOfTU = false);
+                                     bool AtEndOfTU = false,
+                                     bool DeducingReturnType = false);
   VarTemplateSpecializationDecl *BuildVarTemplateInstantiation(
       VarTemplateDecl *VarTemplate, VarDecl *FromVar,
       const TemplateArgumentList *PartialSpecArgs,
diff --git a/clang/lib/Sema/SemaTemplateDeduction.cpp b/clang/lib/Sema/SemaTemplateDeduction.cpp
index acd1151184e4..b9918c556443 100644
--- a/clang/lib/Sema/SemaTemplateDeduction.cpp
+++ b/clang/lib/Sema/SemaTemplateDeduction.cpp
@@ -5459,7 +5459,7 @@ bool Sema::DeduceReturnType(FunctionDecl *FD, SourceLocation Loc,
 
   if (FD->getTemplateInstantiationPattern()) {
     runWithSufficientStackSpace(Loc, [&] {
-      InstantiateFunctionDefinition(Loc, FD);
+      InstantiateFunctionDefinition(Loc, FD, false, false, false, /*DeducingReturnType=*/true);
     });
   }
 
diff --git a/clang/lib/Sema/SemaTemplateInstantiateDecl.cpp b/clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
index e058afe81da5..e3d34c305429 100644
--- a/clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
+++ b/clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
@@ -4955,7 +4955,8 @@ void Sema::InstantiateFunctionDefinition(SourceLocation PointOfInstantiation,
                                          FunctionDecl *Function,
                                          bool Recursive,
                                          bool DefinitionRequired,
-                                         bool AtEndOfTU) {
+                                         bool AtEndOfTU,
+                                         bool DeducingReturnType) {
   if (Function->isInvalidDecl() || isa<CXXDeductionGuideDecl>(Function))
     return;
 
@@ -5284,8 +5285,11 @@ void Sema::InstantiateFunctionDefinition(SourceLocation PointOfInstantiation,
     savedContext.pop();
   }
 
-  DeclGroupRef DG(Function);
-  Consumer.HandleTopLevelDecl(DG);
+  if(!DeducingReturnType || !(parentEvaluationContext().isUnevaluated()
+       || parentEvaluationContext().isImmediateFunctionContext())) {
+     DeclGroupRef DG(Function);
+     Consumer.HandleTopLevelDecl(DG);
+  }
 
   // This class may have local implicit instantiations that need to be
   // instantiation within this scope.
```


The `DeducingReturnType` boolean is only useful not to break 
`CodeGenCXX/template-param-objects-address-space.cpp` so maybe we'd need to investigate that test further
 


@erichkeane @zygoloid @AaronBallman

erichkeane commented:
That seems like a heavy hand.  Should we instead be checking the 'unevaluated context' state to see whether we should be emitting stuff?  I can't imagine that anything from an unevaluated context should ever be emitted.
